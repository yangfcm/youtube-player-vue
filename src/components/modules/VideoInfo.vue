<template>
  <div>
    <app-blank></app-blank>
    <div class="ui unstackable items">
      <div class="item">
        <div class="content">
          <h2 class="ui blue header">
            {{ videoDetail.snippet.title }}
          </h2>
          <div style="margin-bottom: 8px;"></div>
          <router-link :to="`/channel/${videoDetail.snippet.channelId}`">
            {{ videoDetail.snippet.channelTitle }}
          </router-link>
          <div style="margin-bottom: 12px;"></div>
          <div class="meta">
            <span>
              <i class="calendar plus icon"> </i>
              {{ videoDetail.snippet.publishedAt | date }}
              <span style="margin-right: 8px;"></span>
              <i class="eye icon"></i>
              {{ videoDetail.statistics.viewCount | decimal }}
            </span>
            <span class="ui right floated">
              <i class="thumbs up icon"></i>
              {{ videoDetail.statistics.likeCount | decimal }}
              <span style="margin-right: 8px;"></span>
              <i class="thumbs down icon"></i>
              {{ videoDetail.statistics.dislikeCount | decimal }}
            </span>
          </div>
          <app-breakline></app-breakline>
          <div class="description" style="overflow: auto;">
            {{ videoDetail.snippet.description }}
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import Blank from "../common/Blank";
import Breakline from "../common/Breakline";

export default {
  components: {
    appBlank: Blank,
    appBreakline: Breakline,
  },
  props: ["videoDetail"],
};
</script>

<style scoped></style>
