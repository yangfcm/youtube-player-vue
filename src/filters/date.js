/**
 * Convert a date object to human-readable string
 */
export default (date) => {
  const dateObj = new Date(date);
  const day = dateObj.getDate();
  const month = new Intl.DateTimeFormat("en-US", {
    month: "short",
  }).format(dateObj);
  const year = dateObj.getFullYear();
  return `${day} ${month}, ${year}`;
};
