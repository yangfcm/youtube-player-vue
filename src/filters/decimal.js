/**
 * Separate an integer number by commas
 * e.g. 12345678 => 12,345,678
 */
export default (num) => {
  if (isNaN(Number(num))) return "";
  let strNum = String(num);
  let arrNum = strNum.split("");
  let arr = [];
  let part = "";
  for (let i = 0; i < arrNum.length; i++) {
    let num = arrNum[arrNum.length - i - 1];
    part = num + part;
    if ((i + 1) % 3 === 0) {
      arr.unshift(part);
      part = "";
    }
  }
  if (part.length > 0) {
    arr.unshift(part);
  }
  return arr.join(",");
};
