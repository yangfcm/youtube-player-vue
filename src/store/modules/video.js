import axios from "../../settings";
import { FETCH_VIDEOS, CATCH_VIDEO_ERROR, FETCH_VIDEO } from "../types";
import parseError from "../helpers/parseError";

const state = {
  videos: null,
  video: null,
  videoError: null,
};

const getters = {
  videoErrorMessage: (state) => {
    return parseError(state.videoError, "Failed to fetch video");
  },
};

const mutations = {
  FETCH_VIDEOS: (state, payload) => {
    state.videos = payload;
    state.videoError = null;
  },
  FETCH_VIDEO: (state, payload) => {
    state.video = payload;
    state.videoError = null;
  },
  CATCH_VIDEO_ERROR: (state, payload) => {
    state.videoError = payload;
  },
};

const actions = {
  /** Fetch a list of videos by filter */
  fetchVideos: async (context, filter, pageToken) => {
    try {
      const response = await axios.get("/videos", {
        params: {
          ...axios.defaults.params,
          ...filter,
          part: "snippet,statistics",
          maxResults: 15,
          pageToken,
        },
      });
      context.commit(FETCH_VIDEOS, response.data);
    } catch (err) {
      context.commit(CATCH_VIDEO_ERROR, err);
    }
  },

  /** Fetch a video by id */
  fetchVideo: async (context, videoId) => {
    try {
      const response = await axios.get("/videos", {
        params: {
          ...axios.defaults.params,
          part: "snippet,statistics",
          id: videoId,
        },
      });
      // console.log(response);
      context.commit(FETCH_VIDEO, response.data);
    } catch (err) {
      // console.log("error-==================", e);
      context.commit(CATCH_VIDEO_ERROR, err);
    }
  },
};

export default {
  state,
  getters,
  mutations,
  actions,
};
