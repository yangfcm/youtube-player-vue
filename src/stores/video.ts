import { ref, computed } from 'vue'
import { defineStore } from 'pinia'
import { AsyncStatus } from '@/settings/types'
import { type VideoMetaSnippetStats, type VideoResponse } from './types'
import { fetchVideoAPI } from './api'
import { type AxiosResponse } from 'axios'

type VideoStore = {
  error: string
  status: AsyncStatus
  video: Record<string, VideoMetaSnippetStats>
}

export const useVideoStore = defineStore('video', () => {
  const videoStore = ref<VideoStore>({
    error: '',
    status: AsyncStatus.IDLE,
    video: {},
  })

  const status = computed(() => videoStore.value.status)
  const error = computed(() => videoStore.value.error)
  const video = computed(() => videoStore.value.video || {})

  const fetchVideo = async (videoId: string) => {
    try {
      videoStore.value.status = AsyncStatus.LOADING
      videoStore.value.error = ''
      const response: AxiosResponse<VideoResponse> = await fetchVideoAPI(videoId)
      const video = response.data.items[0]
      if (video) {
        videoStore.value.video = {
          ...(videoStore.value.video || {}),
          [video.id as string]: video,
        }
      }
      videoStore.value.status = AsyncStatus.SUCCESS
    } catch (err: any) {
      videoStore.value.status = AsyncStatus.FAIL
      videoStore.value.error = err.message
    }
  }

  return {
    videoState: videoStore,
    video,
    status,
    error,
    fetchVideo,
  }
})
