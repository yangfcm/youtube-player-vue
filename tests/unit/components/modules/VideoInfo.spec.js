import { shallowMount, RouterLinkStub, createLocalVue } from "@vue/test-utils";
import VideoInfo from "@/components/modules/VideoInfo";
import { videoItem1 as video } from "../../store/fixtures/video";
import date from "@/filters/date";
import decimal from "@/filters/decimal";

describe("Test VideoInfo component", () => {
  const localVue = createLocalVue();
  localVue.filter("date", date);
  localVue.filter("decimal", decimal);
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(VideoInfo, {
      propsData: {
        videoDetail: video,
      },
      stubs: {
        RouterLink: RouterLinkStub,
      },
      localVue,
    });
  });
  afterEach(() => {
    wrapper.destroy();
  });

  it("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render correct content", () => {
    expect(wrapper.find("h2").text()).toContain(video.snippet.title);
    expect(wrapper.find("a").props().to).toBe(
      `/channel/${video.snippet.channelId}`
    );
    expect(wrapper.find("div.description").text()).toContain(
      video.snippet.description
    );
  });
});
