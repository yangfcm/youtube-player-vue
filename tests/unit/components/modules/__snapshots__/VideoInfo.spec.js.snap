// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test VideoInfo component should match snapshot 1`] = `
<div>
  <app-blank-stub></app-blank-stub>
  <div class="ui unstackable items">
    <div class="item">
      <div class="content">
        <h2 class="ui blue header">
          Mock video title item1
        </h2>
        <div style="margin-bottom: 8px;"></div> <a>
          Ford Motor Company
        </a>
        <div style="margin-bottom: 12px;"></div>
        <div class="meta"><span><i class="calendar plus icon"></i>
            14 Jul, 2020
            <span style="margin-right: 8px;"></span> <i class="eye icon"></i>
          1,766,115
          </span> <span class="ui right floated"><i class="thumbs up icon"></i>
            36,139
            <span style="margin-right: 8px;"></span> <i class="thumbs down icon"></i>
          1,372
          </span></div>
        <app-breakline-stub></app-breakline-stub>
        <div class="description" style="overflow: auto;">Mock channel description item1</div>
      </div>
    </div>
  </div>
</div>
`;
